extension/k8sleaderelector/extension.go

```diff
@@ -36,17 +36,62 @@ type leaderElectionExtension struct {
 	client        kubernetes.Interface
 	logger        *zap.Logger
 	leaseHolderID string
+	cancel        context.CancelFunc
 
 	onStartedLeading []StartCallback
 	onStoppedLeading []StopCallback
 }
 
+// If the receiver sets a callback function then it would be invoked when the leader wins the election
+func (lee *leaderElectionExtension) startedLeading(ctx context.Context) {
+	for _, callback := range lee.onStartedLeading {
+		callback(ctx)
+	}
+}
+
+// If the receiver sets a callback function then it would be invoked when the leader loss the election
+func (lee *leaderElectionExtension) stoppedLeading() {
+	for _, callback := range lee.onStoppedLeading {
+		callback()
+	}
+}
+
 // Start begins the extension's processing.
 func (lee *leaderElectionExtension) Start(_ context.Context, _ component.Host) error {
+	lee.logger.Info("Starting k8s leader elector with UUID", zap.String("UUID", lee.leaseHolderID))
+
+	ctx := context.Background()
+	ctx, lee.cancel = context.WithCancel(ctx)
+
+	// Create the K8s leader elector
+	leaderElector, err := newK8sLeaderElector(lee.config, lee.client, lee.startedLeading, lee.stoppedLeading, lee.leaseHolderID)
+	if err != nil {
+		lee.logger.Error("Failed to create k8s leader elector", zap.Error(err))
+		return err
+	}
+
+	go func() {
+		// Leader election loop stops if context is canceled or the leader elector loses the lease.
+		// The loop allows continued participation in leader election, even if the lease is lost.
+		for {
+			leaderElector.Run(ctx)
+
+			if ctx.Err() != nil {
+				break
```

Comment:
I believe we need to add a synchronization here to release `leaderElectionExtension.Shutdown` call which should wait for this.

Reply:
fixed.